File: frontend/utils/chunker/table/linearizeTable.ts
/** Parse a single markdown table row (pipe‑delimited), ignoring leading /
 *  trailing pipes, returning an array of trimmed cell values. */
function parseTableRow(row: string): string[] {
  const cleaned = row.replace(/^\|/, "").replace(/\|$/, "");
  return cleaned.split("|").map((cell) => cell.trim());
}

/** True when the row consists only of pipes, spaces, hyphens and colons
 *  and contains at least three hyphens — a markdown table separator.
 *  Handles all alignment variants: |---|, |:---|, |---:|, |:---:|. */
function isSeparatorRow(row: string): boolean {
  const trimmed = row.trim();
  return /^[\|\s\-:]+$/.test(trimmed) && /:?-{3,}:?/.test(trimmed);
}

/**
 * Convert a raw markdown table block into an array of natural‑language
 * sentences, one per data row.
 *
 * Example:
 * ```
 * | Particle | Charge | Location  |
 * |----------|--------|-----------|
 * | Proton   | +1     | Nucleus   |
 * | Electron | -1     | Orbital   |
 * ```
 * →
 * ```
 * "1. Particle: Proton. Charge: +1. Location: Nucleus"
 * "2. Particle: Electron. Charge: -1. Location: Orbital"
 * ```
 *
 * Each output row is prefixed with its 1‑based index for positional
 * awareness in embedding / LLM contexts.
 */
export function linearizeTable(tableText: string): string[] {
  const lines = tableText.split("\n").filter((l) => l.trim().length > 0);
  if (lines.length < 2) return [tableText];

  // Locate the separator row (|---|).
  let headerLine = "";
  let separatorIndex = -1;
  for (let i = 0; i < lines.length; i++) {
    if (isSeparatorRow(lines[i])) {
      separatorIndex = i;
      break;
    }
  }

  if (separatorIndex === -1) return [tableText];

  // Header row is the line immediately before the separator.
  headerLine = lines[separatorIndex - 1] || "";
  const headers = parseTableRow(headerLine);

  // All lines after the separator are data rows.
  const dataRows = lines.slice(separatorIndex + 1);
  const result: string[] = [];

  for (const row of dataRows) {
    const cells = parseTableRow(row);
    if (cells.length === 0) continue;

    // Build sentence: "header: value. header: value. …"
    const parts: string[] = [];
    for (let i = 0; i < headers.length && i < cells.length; i++) {
      const header = headers[i];
      const value = cells[i];
      if (value === "" || value === undefined) continue;
      parts.push(`${header}: ${value}`);
    }
    if (parts.length > 0) {
      result.push(parts.join(". "));
    }
  }

  // Prepend 1‑based row index so the chunk knows its position in the table.
  const numbered = result.map((desc, idx) => `${idx + 1}. ${desc}`);
  return numbered.length > 0 ? numbered : [tableText];
}

File: src-tauri/src/chunker/table/linearize_table.rs
use regex::Regex;
use std::sync::LazyLock;

static RE_CHARS: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^[|\s\-:]+$").unwrap());
static RE_HYPHENS: LazyLock<Regex> = LazyLock::new(|| Regex::new(r":?-{3,}:?").unwrap());

fn parse_table_row(row: &str) -> Vec<String> {
    let mut cleaned = row.trim();
    if cleaned.starts_with('|') {
        cleaned = &cleaned[1..];
    }
    if cleaned.ends_with('|') {
        cleaned = &cleaned[..cleaned.len() - 1];
    }
    cleaned.split('|').map(|cell| cell.trim().to_string()).collect()
}

fn is_separator_row(row: &str) -> bool {
    let row_trim = row.trim();
    if !row_trim.starts_with('|') || !row_trim.ends_with('|') {
        return false;
    }
    
    if !RE_CHARS.is_match(row_trim) {
        return false;
    }
    
    let parts: Vec<&str> = row_trim.split('|').filter(|s| !s.is_empty()).collect();
    if parts.is_empty() {
        return false;
    }
    
    parts.iter().all(|p| RE_HYPHENS.is_match(p.trim()))
}

pub fn linearize_table(table_text: &str) -> Vec<String> {
    let lines: Vec<&str> = table_text.split('\n').filter(|l| !l.trim().is_empty()).collect();
    if lines.len() < 2 {
        return vec![table_text.to_string()];
    }

    let mut separator_index = None;
    for (idx, line) in lines.iter().enumerate() {
        if is_separator_row(line) {
            separator_index = Some(idx);
            break;
        }
    }

    let sep_idx = match separator_index {
        Some(idx) => idx,
        None => return vec![table_text.to_string()],
    };

    if sep_idx == 0 {
        return vec![table_text.to_string()];
    }

    let header_line = lines[sep_idx - 1];
    let headers = parse_table_row(header_line);

    let data_rows = &lines[sep_idx + 1..];
    let mut result = Vec::new();

    for row in data_rows {
        let cells = parse_table_row(row);
        if cells.is_empty() {
            continue;
        }

        let mut parts = Vec::new();
        for i in 0..headers.len() {
            if i < cells.len() {
                let header = &headers[i];
                let value = &cells[i];
                if !value.is_empty() {
                    parts.push(format!("{}: {}", header, value));
                }
            }
        }

        if !parts.is_empty() {
            result.push(parts.join(". "));
        }
    }

    let numbered: Vec<String> = result
        .into_iter()
        .enumerate()
        .map(|(idx, desc)| format!("{}. {}", idx + 1, desc))
        .collect();

    if numbered.is_empty() {
        vec![table_text.to_string()]
    } else {
        numbered
    }
}
